Simplify and cleanup yaml IO

QRemy opened the pull request:
Simplify and cleanup yaml IO:
-use more the functions in utlis so we don't rewrite the code related to checksum, overwrite, log,  (and maybe meta data creator? )...
-simplify FitResult. write to use directly Models.write with an extra_dict (so we don't need #5311)

bkhelifi commented:
This feature looks nice... can you finish the PR such that all the tests are OK? Take care of the coverage. Thanks @QRemy

Astro-Kirsty reviewed (approved):
Thanks @QRemy 
These are some nice adaptations

bkhelifi reviewed (approved):
Thanks @QRemy. It looks OK for me ;)

bkhelifi commented on gammapy/analysis/config.py:
```diff
@@ -255,19 +256,12 @@ def from_yaml(cls, config_str):
 
     def write(self, path, overwrite=False):
```
Can you profit of this PR to detail the arguments of the function?

bkhelifi commented on gammapy/modeling/fit.py:
```diff
@@ -767,37 +765,13 @@ def covariance_result(self):
         path,
         overwrite=False,
-        full_output=False,
+        full_output=True,
```
The docstring should be updated ;)

bkhelifi commented on gammapy/modeling/models/core.py:
```diff
@@ -518,6 +518,8 @@ def write(
             When True adds a CHECKSUM entry to the file.
             Default is False.
+        extra_dict : dict
+            Additionnal informations to write with the models.
```
Precise the default

QRemy replied:
removed from public API

bkhelifi commented on gammapy/utils/scripts.py:
```diff
@@ -67,8 +68,8 @@ def read_yaml(filename, logger=None, checksum=False):
 
-def write_yaml(dictionary, filename, logger=None, sort_keys=True, checksum=False):
-    """Write YAML file.
+def to_yaml(dictionary, sort_keys=False):
```
The docstring should be updated because of the change of the parameters...
Should we add an API change decorator?

registerrier replied:
This is a new function here.

registerrier reviewed (commented):
Thanks @QRemy . Yaml serialization is indeed something that needs to be cleaned-up and clarified. 
I have a few questions here:
- I am not totally fond of having models.write() having a possible extra dictionary. This can make the format definition uncertain.
- `write_yaml(to_yaml())` is not very nice.

QRemy commented on gammapy/modeling/fit.py:
```diff
@@ -586,10 +582,12 @@ def _repr_html_(self):
-            "method": self.method,
-            "success": self.success,
-            "message": self.message,
+            self.__class__.__name__: {
```
I introduced that so the to_dict of CovarianceResult and FitResult already write their name rather than having this done within FitResult

QRemy commented on gammapy/modeling/fit.py:
```diff
@@ -767,37 +765,13 @@ def covariance_result(self):
         path,
         overwrite=False,
-        full_output=False,
+        full_output=True,
```
I added this to have also the min/max on the parameters even if they are default because the FitResult should give the full information about the fit conditions.

QRemy commented on gammapy/modeling/fit.py:
```diff
@@ -680,8 +678,8 @@ def __str__(self):
         output = super().to_dict()
-        output["nfev"] = self.nfev
-        output["total_stat"] = self._total_stat
+        output[self.__class__.__name__]["nfev"] = self.nfev
```
this has been introduced after 1.2 so it's fine no ?

QRemy commented on gammapy/utils/scripts.py:
```diff
@@ -67,8 +68,8 @@ def read_yaml(filename, logger=None, checksum=False):
 
-def write_yaml(dictionary, filename, logger=None, sort_keys=True, checksum=False):
-    """Write YAML file.
+def to_yaml(dictionary, sort_keys=False):
```
I introduced the global value just for the cleanup, not sure we want the other options to be configurable

QRemy commented:
> * `write_yaml(to_yaml())` is not very nice.

modified so it looks like :
```
yaml_str = to_yaml()
write_yaml(yaml_str)
```

 I changed the `write_yaml` from taking a dict to a yaml input because it was used almost nowhere as most of the classes relied on their `to_yaml` in their write methods.

QRemy commented:
> * I am not totally fond of having models.write() having a possible extra dictionary. This can make the format definition uncertain.

I introduced a private function `_write_models` that is used by both `Models` and `FitResult` so the `extra_dict` mechanism is not exposed.

registerrier reviewed (commented):
Thanks @QRemy. This looks good. 
See the few remaining inline comments. In particular, it misses a `deprecated_argument` decorator for `write_yaml`, as it is exposed in the doc
